oprf/oprf.go

```diff
@@ -67,108 +72,142 @@ type Evaluation struct {
 
 // KeyPair is an struct containing a public and private key.
 type KeyPair struct {
-	publicKey  *group.Element
-	privateKey *group.Scalar
+	publicKey  group.Element
+	privateKey group.Scalar
 }
 
 // Client is a representation of a Client during protocol execution.
 type Client struct {
-	suite   *group.Ciphersuite
+	suite   *suite
 	context []byte
 }
 
 // Server is a representation of a Server during protocol execution.
 type Server struct {
-	suite   *group.Ciphersuite
+	suite   *suite
 	context []byte
 	Kp      KeyPair
 }
 
-func generateContext(id SuiteID) []byte {
-	context := [3]byte{OPRFMode, 0, byte(id)}
+type suite struct {
+	group.Group
+	crypto.Hash
+}
 
-	return context[:]
+func (id SuiteID) String() string {
+	switch id {
+	case OPRFP256:
+		return "OPRFP256"
+	case OPRFP384:
+		return "OPRFP384"
+	case OPRFP521:
+		return "OPRFP521"
+	default:
+		panic(ErrUnsupportedSuite)
+	}
 }
 
 // Serialize serializes a KeyPair elements into byte arrays.
 func (kp *KeyPair) Serialize() []byte {
-	return kp.privateKey.Serialize()
+	data, err := kp.privateKey.MarshalBinary()
+	if err != nil {
+		panic("error on serializing")
+	}
+	return data
 }
 
 // Deserialize deserializes a KeyPair into an element and field element of the group.
-func (kp *KeyPair) Deserialize(suite *group.Ciphersuite, encoded []byte) error {
-	privateKey := group.NewScalar(suite.Curve)
-	privateKey.Deserialize(encoded)
-	publicKey := suite.Generator().ScalarBaseMult(privateKey)
+func (kp *KeyPair) Deserialize(s *suite, encoded []byte) error {
+	privateKey := s.NewScl()
+	err := privateKey.UnmarshalBinary(encoded)
+	if err != nil {
+		return err
+	}
+	publicKey := s.NewElt()
+	publicKey.MulGen(privateKey)
 
 	kp.publicKey = publicKey
 	kp.privateKey = privateKey
 
 	return nil
 }
 
+func generateContext(id SuiteID) (ctx [3]byte) {
+	ctx[0] = OPRFMode
+	ctx[1] = 0
+	ctx[2] = byte(id)
+	return
+}
+
 // GenerateKeyPair generates a KeyPair in accordance with the group.
-func GenerateKeyPair(suite *group.Ciphersuite) KeyPair {
-	privateKey := suite.RandomScalar()
-	publicKey := suite.Generator().ScalarBaseMult(privateKey)
+func GenerateKeyPair(s *suite) KeyPair {
+	r := s.Random()
+	privateKey := r.RndScl(rand.Reader)
```

Comment:
I don't really see why we need to call an interface and then the method itself.

Reply:
This is because I implemented random using randomBytes+hashing.
There maybe other ways to implement random that be not so expensive.

Reply:
refactoring ideas are welcomed.

Reply:
this was refactored